common/src/main/java/io/netty/util/internal/PlatformDependent0.java

```diff
@@ -622,72 +622,57 @@ static void setMemory(Object o, long offset, long bytes, byte value) {
     }
 
     static boolean equals(byte[] bytes1, int startPos1, byte[] bytes2, int startPos2, int length) {
-        if (length <= 0) {
-            return true;
-        }
-        final long baseOffset1 = BYTE_ARRAY_BASE_OFFSET + startPos1;
-        final long baseOffset2 = BYTE_ARRAY_BASE_OFFSET + startPos2;
         int remainingBytes = length & 7;
-        final long end = baseOffset1 + remainingBytes;
-        for (long i = baseOffset1 - 8 + length, j = baseOffset2 - 8 + length; i >= end; i -= 8, j -= 8) {
-            if (UNSAFE.getLong(bytes1, i) != UNSAFE.getLong(bytes2, j)) {
-                return false;
+        final long baseOffset1 = BYTE_ARRAY_BASE_OFFSET + startPos1;
+        final long diff = startPos2 - startPos1;
+        if (length > 7) {
+            final long end = baseOffset1 + remainingBytes;
+            for (long i = baseOffset1 - 8 + length; i >= end; i -= 8) {
```

Comment:
@normanmaurer IIRC Java aligns objects on word boundaries, which would imply arrays also start on word boundaries.  Doesn't this imply that iterating from a likely odd number offset backwards would result in torn reads?  Im surprised this isn't starting at the beginning.

Reply:
good point. @njhill mind checking ?

Reply:
@carl-mastrangelo I just left that aspect as it was before, and assumed the reason in this particular method at least was that statistically strings are more likely to differ near the end than the start (common prefixes). That consideration wouldn't apply to the other methods though which always cover all the bytes.

Also note that this is comparing arbitrary regions so the start might not align with a word boundary either, and even if it did it might not for the destination array. And presumably `BYTE_ARRAY_BASE_OFFSET` would need to be taken into account too.

So I guess it would need some experiments and if there was a big win for aligning the long reads in this way we could consider special-casing very large inputs for the other methods. But I'm not sure how common large inputs are in general since I think this is mainly used for headers?

Reply:
> I think this is mainly used for headers?

Well, at least for gRPC, I think we use the constant time one usually, because we can't tell if a header is sensitive (like an auth header).  

> Also note that this is comparing arbitrary regions so the start might not align with a word boundary either.

Yeah, I considered the same thing, but my gut says pretty much everyone starts at 0.   Since this function is public and not specific to Hpack, we don't know the scope of usage.   You are probably right that it doesn't matter, I was jut expressing surprise for the original structure.